Document pytest and nbval (#306)
MDAnalysis/UserGuide: doc/source/contributing_docs.rst

Thus, cells not only have to work (no errors), but also must give the same output they gave when
they were saved.

To test all notebooks, just cd into ``UserGuide/tests`` and run ``pytest``.
If you want to test a particular notebook, check the the contents of `pytest.ini`, this file
defines flags that you can also pass directly to pytest.
For example, if you wanted to test the `hole2` notebook::

    pytest --nbval --nbval-current-env --nbval-sanitize-with ./sanitize_output.cfg ../doc/source/examples/analysis/polymers_and_membranes/hole2.ipynb
Where ``--nbval`` tells pytest to use nbval to test Jupyter notebooks, ``--nbval-current-env``
to use the currently loaded python environment (make sure you actually loaded your environment)
instead of trying to use the one that was used when the notebook was saved  and
``--nbval-sanitize-with`` to point pytest to a file full of replacement rules like this one
for example::

    regex: (.*B \[0.*B/s\])
    replace: DOWNLOAD
This tells pytest to scan the outputs of all cells and replace the matching string with the word
*DOWNLOAD*. This is called *sanitization*.

.. _`nbval`: https://nbval.readthedocs.io/en/latest/

> sanitization

Reviewer: ```suggestion
Sanitization
```

""""""""""""

> Exactly matching cell outputs between runs is a high bar for testing and tends to give false positives.
> Otherwise correct cells may give different outputs each time they are run (e.g. cells with code

Reviewer: ```suggestion
Exactly matching cell outputs between runs is a high bar for testing and tends to give false errors
-- otherwise correct cells may give different outputs each time they are run (e.g. cells with code
```

I understand what you mean here, but to me a "positive" result when testing would be no tests!

that outputs memory locations).
To alleviate this, before testing each cell pytest will match its output against the regular

> expressions from ``sanitize_output.cfg`` and replace matching variable strings with constant strings
> that won't change between runs and hence prevent spurious failures.

Reviewer: ```suggestion
expressions from ``sanitize_output.cfg``. This file contains replacements for strings that we know will vary.
Pytest will replace the dynamic output with these constant strings, which won't change between runs and hence prevent spurious failures.
```

Just splitting up the sentence a bit!

> If your code correctly outputs variable strings each time its run, you should add a replacement
> rule and try to make it as specific as possible.

Reviewer: ```suggestion
If your code correctly outputs variable strings each time its run, you should add a replacement
rule to the ``sanitize_output.cfg`` file and try to make it as specific as possible.
```

> on the hole2 notebook

Reviewer: ```suggestion
On the hole2 notebook
```

"""""""""""""""""""""

> If you test all notebooks you may run into errors with the *hole2* notebook.
> You can safely ignore them, unless you want to work with it.
> In this case you'll have to download `hole2`_, compile it, and make sure your system can find
> the hole2 executable. In UNIX-based systems this implies adding its path to the ``$PATH``
> environmental variable like this::

Reviewer: ```suggestion
The *hole2* notebook is special in that it requires installation of extra software to run,
namely the `hole2`_ program. If you test all the notebooks you may therefore run
into errors if hole2 is not installed. These errors can be generally ignored unless
you do specifically want to test the hole2 notebook. Of course, you should take
note of other errors that occur if hole2 is installed!
To run the hole2 notebook you'll have to download `hole2`_, compile it, and make sure your system can find
the hole2 executable. In UNIX-based systems this implies adding its path to the ``$PATH``
environmental variable like this::
```

I'd just like to clarify this part a bit more -- the hole2 notebook errors are about the lack of hole2, and if hole2 *is* installed then you should pay attention to the warnings!

Author: Thanks! I applied everything, it's much better now.
